### PR TITLE
Persistent Networks Resources created on host during time of connecti…

### PR DESCRIPTION
**Pull request by shatoboar**
…on, and cleanup during time of removal

### Description

This PR is a GSoC project: https://github.com/apache/cloudstack/issues/4814

This PR adds a feature to L2 persitent networks, where resources of the persistent networks are created at hostConnect phase of DefaultHostListener, and cleaned up at hostAboutToBeRemoved phase. 

This PR also fixes a bug, where rebooting the hosts removes the persistent network's resources, without readding them when the host is readded to the zone.

Fixes: #5196 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
This has been manually tested with a mbx KVM, XenServer and VmWare setup on a local machine. 

Different manual test scenarios:
1. When the persistent network is created and the host is added afterwards, the newly added host has the persistent network's resources.
2. When a host within a zone with a persistent network is removed from the zone, its resources will be cleaned up.
3. When a host with a running VM is removed from the zone, the running VM will migrate before the resources are cleaned up.
4. When a host is rebooted, it still has access to the persistent network's resources.

A Marvin component test under 
test/integration/component/test_persistent_networks.py
added that tests the following steps for a KVM environment.
1. identify hosts in the zone, and remove the first
2. create a L2 persistent network
3. add the host back to the zone

Validation steps:
1. Persistent network state should be implemented before adding the host
2. Host should be added back in successfully
3. Host should have the persistent networks resources after being added

Test output results should be:

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp/MarvinLogs/<LogId>. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
====Trying SSH Connection: Host:172.20.0.121 User:root                                   Port:22 RetryCnt:10===
===SSH to Host 172.20.0.121 port : 22 SUCCESSFUL===
{Cmd: ip addr | grep breth0-991 > /dev/null 2>&1; echo $? via Host: 172.20.0.121} {returns: ['0']}
=== TestName: test_newly_added_host_for_persistent_network_resources | Status : SUCCESS ===

===final results are now copied to: /tmp//MarvinLogs/test_persistent_networks_<LogId>===

### CODE REVIEW

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java [hunk 1]
```diff
@@ -59,12 +70,66 @@
...
+
 
     @Override
     public boolean hostAdded(long hostId) {
         return true;
     }
 
+    private boolean createPersistentNetworkResourcesOnHost(long hostId) {
+        HostVO host = hostDao.findById(hostId);
+        if (host != null) {
```
Comment (GutoVeronezi) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:92]: We could invert this if to reduce indentation and maybe add a log.

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java [hunk 2]
```diff
@@ -59,12 +70,66 @@
...
+                NetworkOfferingVO networkOfferingVO = networkOfferingDao.findById(networkVO.getNetworkOfferingId());
+
+                SetupPersistentNetworkCommand persistentNetworkCommand =
+                        new SetupPersistentNetworkCommand(createNicTOFromNetworkAndOffering(networkVO, networkOfferingVO, host));
+                Answer answer = agentMgr.easySend(hostId, persistentNetworkCommand);
+                if (answer == null) {
+                    throw new CloudRuntimeException("Unable to get answer to the setup persistent network command " + networkVO.getId());
+                }
+                if (!answer.getResult()) {
+                    String msg = "Unable to create L2 persistent network resources from network " + networkVO.getId() + " on the host" + hostId;
```
Comment (GutoVeronezi) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:105]: We could improve this message with more context than only id. Also we could use String.format here.

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java [hunk 3]
```diff
@@ -137,7 +203,26 @@ public boolean hostDisconnected(long hostId, long poolId) {
 
     @Override
     public boolean hostAboutToBeRemoved(long hostId) {
-        return true;
+        // send host the cleanup persistent network resources
+        HostVO host = hostDao.findById(hostId);
+        if (host != null) {
```
Comment (GutoVeronezi) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:208]: We could invert this if to reduce indentation and maybe add a log.

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by Pearl1594** [comment]
Closing this PR - tracking it with PR #5977

**Comment by NuxRo** [comment]
@blueorangutan package

**Pull request closed without merging**